src/main/java/org/hydev/mcpm/client/Downloader.java

```diff
@@ -40,8 +83,45 @@ public void downloadFile(String url, File to)
      */
     public void downloadFiles(Map<String, File> urls, boolean progress, int threads)
     {
-        // TODO: Implement this
-        throw new UnsupportedOperationException("TODO");
+        Thread[] threadss = new Thread[threads];
+        final int UrlPerThread = urls.size() / threads;
+        final int remainingUrl = urls.size() % threads;
+
+        for (int t = 0; t < threads; t++) {
+            final int thread = t;
+            threadss[t] = new Thread() {
+                @Override
+                public void run() {
+                    runThread(urls, threads, thread, UrlPerThread, remainingUrl);
+                }
+            };
+        }
+
+        for (Thread t: threadss) {
+            t.start();
+        }
+    }
+    private void runThread(Map<String, File> urls, int threads, int thread, int UrlPerThread, int remainingUrl) {
+        // Store all urls into a list
+        List<String> files = new ArrayList<>();
+        for (Map.Entry<String, File> url:urls.entrySet()) {
+            files.add(url.getKey());
+        }
```

Comment:
Consider `Set<String> files = urls.keySet()` or if you want a list:

```suggestion
        var files = urls.keySet().stream().toList();
```

Reply:
noted